specs/merkle-tree.md

```diff
@@ -0,0 +1,166 @@
+* `string uri`: An optional field that contains the URI of the deployment. Its purpose is to provide a public record that allows anyone to re-assemble the entire deployment from scratch. This may include the Solidity compiler inputs, which are required for Etherscan verification. The format, location, and contents of the URI are determined by off-chain tooling.
+* `bool arbitraryChain`: If this is `true`, the Merkle root can be executed on any chain without the explicit permission of the Gnosis Safe owners. This is useful if the owners want their system to be permissionlessly deployed on new chains. By default, this is disabled, which means that the Gnosis Safe owners must explicitly approve the deployment on individual chains.
+
+### `EXECUTE` Leaf Data
+
+The `data` field of an `EXECUTE` leaf primarily contains data to forward to the Gnosis Safe. It consists of the following fields, which are ABI encoded:
+* `address to`: The target address of the transaction. This is _not_ the address of the Gnosis Safe.
+* `uint256 value`: The amount of native gas token to transfer from the Gnosis Safe to the target address. This value is not transferred from the `SphinxModuleProxy`.
+* `uint256 gas`: The amount of gas that's included in the call from the `SphinxModuleProxy` to the Gnosis Safe.
+* `bytes txData`: The transaction's data.
+* `Enum.Operation operation`: The type of transaction to execute in the Gnosis Safe, i.e. `Call` or `DelegateCall`.
+* `bool requireSuccess`: If this is `true` and the transaction in the Gnosis Safe fails, the deployment is marked as "failed" and will end immediately. If this is `false`, the deployment will continue regardless of whether the transaction fails.
+
+### `CANCEL` Leaf Data
+
+The `data` field of an `CANCEL` leaf consists of the following fields, which are ABI encoded:
+* `address leafSafeProxy`: See [`APPROVE` leaf data](#approve-leaf-data).
+* `address moduleProxy`: See [`APPROVE` leaf data](#approve-leaf-data).
+* `uint256 leafMerkleRootNonce`: See [`APPROVE` leaf data](#approve-leaf-data).
+* `bytes32 merkleRootToCancel`: The Merkle root of the deployment that should be canceled.
+* `address executor`: See [`APPROVE` leaf data](#approve-leaf-data).
+* `string memory uri`: See [`APPROVE` leaf data](#approve-leaf-data).
+
+
+## High-Level Merkle Tree Invariants
+The purpose of these invariants is to define an unambiguous, consistent structure for Sphinx Merkle trees. This ensures that it's easy for off-chain tooling to recreate the exact Merkle tree signed by the user, and to check that a given Merkle tree is valid. It's worth mentioning that it's possible to construct Merkle trees that do not adhere to this spec, but are executable on-chain. We discuss this possibility and the expected executor behavior in more detail in the [Assumptions section](#assumptions).
+
+### 1. Must be executable on all chains for which there is at least one Merkle leaf.
```

Comment:
The auditors would probably find it helpful if we indicate how we test that this is true, since it's not tested in `merkle-tree.ts`